internal/builtin/providers/tf/provider_functions.go

```diff
@@ -0,0 +1,194 @@
+	}
+	vals := make(map[string]cty.Value)
+	for name, attr := range attrs {
+		val, diag := attr.Expr.Value(nil)
+		if diag.HasErrors() {
+			return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+		}
+		vals[name] = val
+	}
+	return cty.ObjectVal(vals), nil
+}
+
+// encodeTFVarsFunc encodes an object into a string with the same format as a TFVars file
+type encodeTFVarsFunc struct{}
+
+func (f *encodeTFVarsFunc) Name() string {
+	return "encode_tfvars"
+}
+
+func (f *encodeTFVarsFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name: "input",
+			// The input type is determined at runtime
+			Type:              cty.DynamicPseudoType,
+			Description:       "input to encode for TFVars file. Must be an object with key that are valid identifiers",
+			DescriptionFormat: providers.TextFormattingPlain,
+		},
+	}
+	return providers.FunctionSpec{
+		Parameters:        params,
+		Return:            cty.String,
+		Summary:           "encode an object into a string with the same format as a TFVars file",
+		Description:       "provider::terraform::encode_tfvars encodes an object into a string with the same format as a TFVars file",
+		DescriptionFormat: providers.TextFormattingPlain,
+	}
+}
+
+var InvalidInputError = errors.New("invalid input")
+
+func (f *encodeTFVarsFunc) Call(args []cty.Value) (cty.Value, error) {
+	toEncode := args[0]
+	// null is invalid input
+	if toEncode.IsNull() {
+		return cty.NullVal(cty.String), fmt.Errorf("%w: must not be null", InvalidInputError)
+	}
+	if !toEncode.Type().IsObjectType() {
+		return cty.NullVal(cty.String), fmt.Errorf("%w: must be an object", InvalidInputError)
+	}
+	ef := hclwrite.NewEmptyFile()
+	body := ef.Body()
+
+	// Iterate over the elements of the input value
+	it := toEncode.ElementIterator()
+	for it.Next() {
+		key, val := it.Element()
+		// Check if the key is a string, known and not null, otherwise AsString method panics
+		if !key.Type().Equals(cty.String) || !key.IsKnown() || key.IsNull() {
+			return cty.NullVal(cty.String), fmt.Errorf("%w: object key must be a string: %v", InvalidInputError, key) //TODO errors
+		}
+		name := key.AsString()
+		if valid := hclsyntax.ValidIdentifier(name); !valid {
+			return cty.NullVal(cty.String), fmt.Errorf("%w: object key: %s - must be a valid identifier", InvalidInputError, name)
+		}
+		body.SetAttributeValue(key.AsString(), val)
+	}
+	b := ef.Bytes()
+	return cty.StringVal(string(b)), nil
+}
+
+// encodeExprFunc encodes an expression into a string
+type encodeExprFunc struct{}
+
+func (f *encodeExprFunc) Name() string {
+	return "encode_expr"
+}
+
+func (f *encodeExprFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name:              "expr",
+			Type:              cty.DynamicPseudoType,
+			Description:       "expression to encode",
```

Comment:
```suggestion
			Description:       "Expression to encode",
```

Reply:
The word `expression` is still lowercase